hs-bindgen/src-internal/HsBindgen/Frontend/Pass/Parse/Type/Monad.hs

```diff
@@ -36,24 +39,32 @@ import HsBindgen.Util.Tracer
 
   Parsing declarations and parsing types have different requirements, so each
   gets its own monad.
+
+  The ParseType monad includes state for caching typedef resolution to avoid
+  redundant type parsing during typedef chain resolution. It is parametrized
+  over the phase type to avoid circular dependencies.
 -------------------------------------------------------------------------------}
 
-newtype ParseType a = Wrap {
-      unwrap :: IO a
+newtype ParseType p a = Wrap {
+      unwrap :: StateT (Map CXType (C.Type p)) IO a
     }
   deriving newtype (
       Functor
     , Applicative
     , Monad
     , MonadIO
+    , MonadState (Map CXType (C.Type p))
```

Comment:
Don't derive it and offer an API for caching instead

Reply:
But this still needs to be done.